### ISSUE TITLE
Support for declaring out field of OnResolveResult in plugin?

### ISSUE THREAD

**Issue by noyobo** [evanw/esbuild#3133]
The file is analyzed in the plugin at build runtime, and I want to specify the output path of the file, like

```js
build.onResolve({ filter: /\.*/, namespace: 'file' }, async (args) => {
    if (isComponent(args.path)) {
		return  {
			path: args.path,
			out: 'dist/components/..'
		}
    }
});
```

This is very helpful for me to port the webpack build program, which added dependencies at runtime through the `SingleEntryPlugin` in the previous webpack build program.

**Comment by evanw** [comment]
This proposed API doesn't make sense for multiple reasons. One reason is that `onResolve` can be called multiple times for a given input file, so this would be ill-defined if each `onResolve` call returned a different value (especially because these are called in parallel in a non-deterministic order). So `onLoad` makes more sense than `onResolve` since `onLoad` is only called once per input file. However, you are asking to be able to set the path of an output file, not an input file. So using `onLoad` doesn't make sense either because that operates on input files, not output files. Since esbuild is a bundler, there isn't a 1:1 mapping between input files and output files.

If you want to set the output path of an entry point, you can do that when you specify the entry point: https://esbuild.github.io/api/#entry-points. Alternatively if you're asking to set the output path of other files (including intermediate code splitting chunks), that is already tracked by #553. In any case I'm closing this issue because I do not plan to make the proposed API change.

**Comment by noyobo** [comment]
In this case, there must be explicit entryPoints. Due to the changing dependencies of some files based on their content, it is necessary to analyze the files and obtain clear entryPoints before each `esbuild.build()` operation. This requirement cannot be fulfilled within `esbuildContext.watch()`.

```js
// app.js
export function App() {
  return <div>app</div>
}
```

entryPoints is  `['app.js']` 


When the content of the file changes: 
```js
// app.js
import Foo from './foo.component.js';
export function App() {
  return <div>app</div>
}
```

must reset entryPoints to  `['app.js', 'foo.component.js']` and call again `esbuild.build()` .

I said earlier that implementing the out field in onResolveResult or onLoadResult may not be a good solution. So what better way to support such a demand?   @evanw

**Comment by evanw** [comment]
If you want to change the entry points, then you can change them and restart esbuild's watch mode on a new esbuild context. This cannot be done from within a plugin because each plugin is specific to a single esbuild context.
